From: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
Date: Thu, 21 Mar 2024 16:52:34 +0800
Subject: [PATCH] wifi: mt76: mt7915: set channel after sta is associated

When sta is associated to an AP operating on a DFS channel, a channel
setting operation is required to change the channel switch reason
from CH_SWITCH_DFS to CH_SWITCH_NORMAL. The switch reason during
authentication is CH_SWITCH_DFS because the DFS state is still
USABLE at that point.

Signed-off-by: StanleyYP Wang <StanleyYP.Wang@mediatek.com>
---
 mt7915/main.c |   28 ++++++++++++++++++++++++++++
 1 files changed, 28 insertions(+), 0 deletions(-)

--- a/mt7915/main.c
+++ b/mt7915/main.c
@@ -780,6 +780,28 @@ mt7915_post_channel_switch(struct ieee80
 	return ret;
 }
 
+static void
+mt7915_event_callback(struct ieee80211_hw *hw, struct ieee80211_vif *vif,
+		      const struct ieee80211_event *event)
+{
+	struct mt7915_phy *phy = mt7915_hw_phy(hw);
+	struct mt7915_dev *dev = phy->dev;
+
+	switch (event->type) {
+	case MLME_EVENT:
+		if (event->u.mlme.data == ASSOC_EVENT &&
+		    event->u.mlme.status == MLME_SUCCESS) {
+			dev_info(dev->mt76.dev, "%s: set channel after association", __func__);
+			ieee80211_stop_queues(hw);
+			mt76_update_channel(phy->mt76);
+			ieee80211_wake_queues(hw);
+		}
+		break;
+	default:
+		break;
+	}
+}
+
 int mt7915_mac_sta_add(struct mt76_dev *mdev, struct ieee80211_vif *vif,
 		       struct ieee80211_sta *sta)
 {
@@ -1907,6 +1929,7 @@ const struct ieee80211_ops mt7915_ops =
 	.set_sar_specs = mt7915_set_sar_specs,
 	.channel_switch_beacon = mt7915_channel_switch_beacon,
 	.post_channel_switch = mt7915_post_channel_switch,
+	.event_callback = mt7915_event_callback,
 	.get_stats = mt7915_get_stats,
 	.get_et_sset_count = mt7915_get_et_sset_count,
 	.get_et_stats = mt7915_get_et_stats,
